orchestrator/orchestrator/src/valset_relaying.rs

```diff
@@ -0,0 +1,134 @@
+    let peggy_id = get_peggy_id(contract_address, our_ethereum_address, web3)
+        .await
+        .expect("Failed to get Peggy ID");
+
+    // confirm arbitrarily many valsets
+    let mut last_unsigned_valset = get_oldest_unsigned_valset(contact, our_cosmos_address).await;
+    while let Ok(valset) = last_unsigned_valset {
+        send_valset_confirm(
+            contact,
+            ethereum_key,
+            Coin {
+                denom: pay_fees_in.clone(),
+                amount: 1u32.into(),
+            },
+            valset.result,
+            cosmos_key,
+            String::from_utf8(peggy_id.clone()).expect("Invalid PeggyID"),
+            None,
+            None,
+            None,
+        )
+        .await
+        .unwrap();
+
+        last_unsigned_valset = get_oldest_unsigned_valset(contact, our_cosmos_address).await;
+    }
+
+    // now that we have caught up on valset requests we should determine if we need to relay one
+    // to Ethereum for that we will find the latest confirmed valset and compare it to the ethereum chain
+    let latest_valsets = get_last_valset_requests(contact).await;
+    if latest_valsets.is_err() {
+        // there are no latest valsets to check, possible on a bootstrapping chain maybe handle better?
+        return;
+    }
+    let latest_valsets = latest_valsets.unwrap();
+
+    let mut latest_confirmed = None;
+    let mut latest_valset = None;
+    info!("Retrieving validator set signatures from the Cosmos chain");
+    for set in latest_valsets.result {
+        let confirms = get_all_valset_confirms(contact, set.nonce).await;
+        if let Ok(confirms) = confirms {
+            latest_confirmed = Some(confirms);
+            latest_valset = Some(set);
+            break;
+        }
+    }
+
+    if latest_confirmed.is_none() {
+        return;
+    }
+    let latest_confirmed = latest_confirmed.unwrap();
+    let latest_valset = latest_valset.unwrap();
+
+    let latest_ethereum_valset = get_valset_nonce(contract_address, our_ethereum_address, web3)
+        .await
+        .expect("Failed to get Ethereum valset");
+    let latest_cosmos_valset: Uint256 = latest_confirmed.height.into();
+    if latest_cosmos_valset > latest_ethereum_valset {
+        info!(
+            "We have detected latest valset {} but latest on Ethereum is {} sending an update!",
+            latest_cosmos_valset, latest_ethereum_valset
+        );
+
+        let old_valset = if latest_ethereum_valset == 0u8.into() {
+            // we need to have a special case for validator set zero, that valset was never stored on chain
+            let mut valset = get_peggy_valset(contact)
+                .await
+                .expect("Failed to get old valset");
+            // make sure the nonce here is zero
+            valset.result.nonce = 0;
+            valset
+        } else {
+            // get the old valset from the Cosmos chain
+            get_peggy_valset_request(contact, latest_ethereum_valset)
```

Comment:
The result contains the "raw" snapshot of orchestrator address/power data at the time of creation. It may include empty Ethereum addresses/ power. Order is by power DESC. 
If you use the data here, I can also "prune" it before returning. Please assign an issue to me if this makes sense.

Note: With the fork prevention delay the cosmos side may not know about any previous in flight multisig updates. De factor we have the the "optimistic" update workflow implemented. If you prefer the pessimistic one then you can delay any updates until `peggy/lastNonce/bridge_multisig_update` returns you a value == `latest_ethereum_valset`.

Reply:
After discussing with Jehan we don't actually need to do any pessimistic locking and can be optimistic. This has to do with a very specific edge case in the contract, validators are signing over the hash of the *next* valset the previous valset which you submit was signed and had it's hash stored in the previous update. Meaning you can have powers change and not invalidate the validator set update as long as the target update doesn't also change. 

It's nuanced and @jtremback  should have an issue up about it at some point.

But getting back to the exact code at hand. Yes what we really need here is to finish the bootstrapping process enough and then look at what the bootstrapped valset is to handle this edgecase of valset nonce 0.